src/client/mapblock_mesh.cpp

```diff
@@ -664,19 +664,22 @@ static u8 face_contents(content_t m1, content_t m2, bool *equivalent,
 	else if (c2 == 0)
 		c2 = f2.visual_solidness;
 
+	// In all cases, only render face if it is opaque.
+	// All transparency is handled by MapblockMeshGenerator.
+
 	if (c1 == c2) {
 		*equivalent = true;
 		// If same solidness, liquid takes precense
-		if (f1.isLiquid())
-			return 1;
-		if (f2.isLiquid())
-			return 2;
+		if (f1.drawtype == NDT_LIQUID)
+			return f1.alpha == ALPHAMODE_OPAQUE ? 1 : 0;
+		if (f2.drawtype == NDT_LIQUID)
+			return f2.alpha == ALPHAMODE_OPAQUE ? 2 : 0;
 	}
 
 	if (c1 > c2)
-		return 1;
+		return f1.drawtype == NDT_NORMAL || f1.drawtype == NDT_PLANTLIKE_ROOTED || f1.alpha == ALPHAMODE_OPAQUE ? 1 : 0;
```

Comment:
This looks like the wrong approach to me, `nodedef.cpp` is supposed to set solidness so that MapBlockMesh _doesn't_ have to do these comparisons.

Reply:
We need a new system for solidness weights then, what is the best place to discuss it?

Reply:
I've implemented new levels of solidness and removed visual_solidness entirely.